Library/Source/XrPlugin.cpp

```diff
@@ -0,0 +1,215 @@
+        static void Initialize(Napi::Env&);
+
+        XrPlugin::XrPlugin(const Napi::CallbackInfo& info);
+        ~XrPlugin();
+
+    private:
+        static Napi::FunctionReference constructor;
+
+        void BeginSession(const Napi::CallbackInfo&); // TODO: Make this asynchronous.
+        void EndSession(const Napi::CallbackInfo&); // TODO: Make this asynchronous.
+        void EndSession();
+        bool IsSessionActive() const;
+        void BeginFrame(const Napi::CallbackInfo&);
+        void EndFrame(const Napi::CallbackInfo&);
+        void EndFrame();
+        Napi::Value GetActiveFrameBuffers(const Napi::CallbackInfo& info);
+
+        xr::System m_hmd{};
+        std::unique_ptr<xr::System::Session> m_session{};
+        std::unique_ptr<xr::System::Session::Frame> m_frame{};
+        FrameBufferManager& m_frameBufferManager;
+        std::vector<FrameBufferData*> m_activeFrameBuffers{};
+
+        std::map<uintptr_t, std::unique_ptr<FrameBufferData>> m_texturesToFrameBuffers{};
+    };
+
+    Napi::FunctionReference XrPlugin::constructor;
+
+    void XrPlugin::Initialize(Napi::Env& env)
+    {
+        Napi::HandleScope scope{ env };
+
+        Napi::Function func = DefineClass(
+            env,
+            "XrPlugin",
+            {
+                InstanceMethod("beginSession", &XrPlugin::BeginSession),
+                InstanceMethod("endSession", &XrPlugin::EndSession),
+                InstanceMethod("beginFrame", &XrPlugin::BeginFrame),
+                InstanceMethod("endFrame", &XrPlugin::EndFrame),
+                InstanceMethod("getActiveFrameBuffers", &XrPlugin::GetActiveFrameBuffers)
+            },
+            nullptr);
+
+        constructor = Napi::Persistent(func);
+        constructor.SuppressDestruct();
+
+        env.Global().Set("XrPlugin", func);
+    }
+
+    XrPlugin::XrPlugin(const Napi::CallbackInfo& info)
+        : Napi::ObjectWrap<XrPlugin>{ info }
+        , m_frameBufferManager{ info[0].As<Napi::External<NativeEngine::Impl>>().Data()->GetFrameBufferManager() }
+    {}
+
+    XrPlugin::~XrPlugin()
+    {
+        if (IsSessionActive())
+        {
+            if (m_frame != nullptr)
+            {
+                EndFrame();
+            }
+
+            EndSession();
+        }
+    }
+
+    // TODO: Make this asynchronous.
+    void XrPlugin::BeginSession(const Napi::CallbackInfo&)
+    {
+        assert(!IsSessionActive());
+        assert(m_frame == nullptr);
+
+        if (!m_hmd.IsInitialized())
+        {
+            while (!m_hmd.TryInitialize());
+        }
+
+        m_session = m_hmd.CreateSession(bgfx::getInternalData()->context);
+    }
+
+    // TODO: Make this asynchronous.
+    void XrPlugin::EndSession(const Napi::CallbackInfo&)
+    {
+        EndSession();
+    }
+
+    void XrPlugin::EndSession()
+    {
+        assert(IsSessionActive());
+        assert(m_frame == nullptr);
+
+        m_session->RequestEndSession();
+
+        while (m_session->GetNextFrame() != nullptr);
```

Comment:
After recent changes to Frame, this will actually never terminate because GetNextFrame() does not return nullptr anymore. Check the provided boolean values instead.